### PR TITLE
Speed up todo output if there is nothing to do

### PR DESCRIPTION
**Pull request by mitchellwrosen**
This patch speeds up `todo` output when there is nothing to do.

- `showTodoOutput` now takes an action that computes a `PrettyPrintEnv` rather than a `PrettyPrintEnv` directly (because it's slow to compute, and only needed if there are things to do)
- `NothingTodo` output message was reused for this output case (rather than making a new `NoTodoOutput` or similar)
- Calculations for whether or not a `TodoOutput` has any conflicts/edits were moved to the `TodoOutput` so they could be called from more than one place.

Maybe mostly addresses https://github.com/unisonweb/unison/issues/1029?

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/Editor/HandleInput.hs [hunk 1]
```diff
@@ -472,7 +472,7 @@ loop = do
           if b then do
             respond (ShowDiff input (Branch.namesDiff destb merged))
             patch <- getPatchAt defaultPatchPath
-            void $ propagatePatch inputDescription patch dest
+            void $ propagatePatch input inputDescription patch dest
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Codebase/Editor/HandleInput.hs]: `input` was piped around here and elsewhere for displaying the `NothingTodo` case.

**Comment by mitchellwrosen** [comment]
Happy consequence: `update`, which also seems to call `todo` for you at the end, is now much faster when your edit frontier is empty (which, in my experience, is most of the time)

**Comment by aryairani** [comment]
What do you think about `NoTodo` in general?  It bugs me a little that we have to thread `input` through.  Should `propagatePatch` even generate output messages, or should it be part of the return value?

**Comment by mitchellwrosen** [comment]
I guess I'm not a fan of `NoTodo` as opposed to just having different constructors for different purposes. Agreed that feeding the input to the output, just so the output can switch on it and tweak the display, is not great. Want to see that changed here, or in a different patch?

**Comment by aryairani** [comment]
Let's do it here if you don't mind, and reduce the diff — I think `NoTodo` was a mistake; we can address its old cases in a separate patch.

**Comment by mitchellwrosen** [comment]
@aryairani done!

**Pull request merged**